doc/user_guide/psy_data.rst

```diff
@@ -0,0 +1,302 @@
+the data types provided. The netcdf kernel writer 
+(see :ref:`psyke_netcdf`) uses the following declaration
+with dl_esm_inf::
+
+    generic, public :: PreDeclareVariable => DeclareScalarInteger, &
+                                             DeclareScalarReal,    &
+                                             DeclareFieldDouble
+    ...
+    subroutine DeclareScalarInteger(this, name, value)
+        implicit none
+        class(PSyDataType), intent(inout) :: this
+        character(*), intent(in) :: name
+        integer, intent(in) :: value
+    ...
+    subroutine DeclareFieldDouble(this, name, value)
+        use field_mod, only : r2d_field
+        implicit none
+        class(PSyDataType), intent(inout) :: this
+        character(*), intent(in) :: name
+        type(r2d_field), intent(in) :: value
+    ...
+
+
+``PreEndDeclaration(this)``
++++++++++++++++++++++++++++
+Called once all variables have been declared.
+
+``ProvideVariable(this, name, value)``
+++++++++++++++++++++++++++++++++++++++
+This function is called for each variable to be provided to the
+runtime library. 
+
+``name``:
+  This is the name of the variable as a string.
+
+``value``:
+  This is the actual content of the variable.
+
+The same function ``ProvideVariable`` is called to provide variable
+name and content before and after the user instrumented region.
+Again it is expected that a library using the API will provide
+a generic interface to distinguish between the various possible data
+types, which will be different for each infrastructure library::
+
+    generic, public :: ProvideVariable => WriteScalarInteger, &
+                                          WriteScalarReal,    &
+                                          WriteFieldDouble
+
+``PreEnd(this)``
+++++++++++++++++
+The method ``PreEnd`` is called after all variables before the instrumented
+region have been provided.
+
+``PostStart(this)``
++++++++++++++++++++
+This is the first call after the instrumented region. It does not take
+any parameters, but the static ``PSyDataType`` instance can be used
+to store the name and number of variables if required. This will be
+followed by calls to ``ProvideVariable``, which is described above.
+
+``PostEnd(this)``
++++++++++++++++++
+This function is the last call after an instrumented region. It indicates
+that all variables have been provided.
+ 
+An example of a library using PSyData is included in PSyclone in the
+directory ``.../lib/extract/netcdf``. This library is used to extract
+kernel input- and output-parameter and store them in a NetCDF file.
+
+.. note::
+
+    Note that only the ``PreDataStart`` call takes the module-
+    and region-name as parameters. If these names are required
+    by the profiling library in different calls, they must
+    be stored in the ``PSyData`` object so that they are
+    available when required.
+
+.. _psy_data_transformation:
+
+PSyData Transformation and "PSyData-Node"
```

Comment:
Please use the same formatting for PSyDataNode as you've used e.g. at line 278 below.